DEVELOPMENT-GUIDE.md

```diff
@@ -0,0 +1,166 @@
+
+### *_auto
+Whenever possible, use classes that provide stack-based tracking. These classes have names with the form `*_auto` (ex: `RSA_auto`).
+
+### BigNumObj
+`BigNumObj ` is essentially a `*_auto` object for `BN` resources, but with `BN` specific logic attached to make it easier to use.
+
+### java_ex
+`java_ex` is a C++ exception which represents a Java exception and can be converted to one (immediately prior to returning to Java) by calling `throw_to_java(JNIEnv*)`
+
+### raii_env 
+`raii_env` wraps an instance of `JNIEnv*` and *MUST* always be used (except for calling `java_ex.throw_to_java()`) and *MUST* always be passed by reference.
+It's primary purpose is to check invariants around critical regions and prevent coding mistakes by failing hard if `JNIEnv*` is used incorrectly.
+
+### java_buffer
+Representation of any sequence of bytes passed in from Java (currently either a `DirectByteBuffer` or a `byte[]`). This does not give direct access to the underlying bytes.
+
+### jni_borrow
+Actual representation of bytes from a `java_buffer`. Due to their interactions with the JVM and GC, they should exist for as short a time as possible.
+While a `jni_borrow` exists, ACCP may be in a critial region and thus any use of `raii_env` will result in the process aborting.
+
+### jni_string
+Represents a `String` object from Java and gives access to the UTF-8 encoded contents.
+
+### SecureBuffer
+The `SecureBuffer` represents a fixed-length array of a type (usually `uint8_t`) which will always zero itself upon destruction.
+
+# About JNI (Java Native Interface)
+The [Java Native Interface](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/jniTOC.html) (JNI) is a standard way for standard Java code to interact with native code (such as that written in C or C++).
+The JNI (specifically version 6.0 as supported by JDK8 and linked above) is a core component of ACCP's implementation as it allows us to connect high-performance implementations in C/C++ with callers from Java.
+Unfortunately, JNI development can be tricky and carries with it many correctness, stability, and performance risks if not done properly.
+There are no short-cuts here except to [read the documentation](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/jniTOC.html).
+The most important sections of the official guide are:
+* [Introduction](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/intro.html)
+* [Java Exceptions](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html#java_exceptions)
+* [JNI Types and Data Structures](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/types.html)
+* [Array Operations](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#array_operations)
+
+## How ACCP uses the JNI
+Just as in many other areas of development, is a wide range of acceptable styles for JNI code.
```

Comment:
Missing words? there are a wide range?

Reply:
"*is* a wide range of styles" or "*are* a wide range of styles"? (I'm genuinely curious.)

In the former case, "is" appears to apply to the singular "range". "There is a range of ......".

In the latter case, "are" appears to apply to the plural "styles" where "range of" acts as an adjectival phrase. (I think that's the term). Thus, it could be reworded as "There are *many* styles" or "There are *blue* styles".